controllers/DataSetAdmin.js

```diff
@@ -0,0 +1,271 @@
+          return res.redirect(this.routePrefix)
+        } 
+        set.rowCount = rows.length
+        set.fields = currentFields
+        this.log.debug( "nxus-dataset saveDataUpload dataset " + setId + " fields:", currentFields )
+        return set.save().then( () => {
+          req.flash('info', "Successfully loaded " + rows.length + " data rows")
+          res.redirect(this.routePrefix) //TODO -- route to the edit form for this dataset instead?
+        })
+      })
+    }).catch((e) => {
+      this.log.error( "nxus-dataset saveDataUpload() error loading data rows  ", e, " at: ", e.stack);
+      req.flash('error', "Error processing rows")
+      return res.redirect(this.routePrefix)
+    })
+  }
+
+  /**
+   * Transform the datarow after storage,
+   * associating with the given dataset ID.
+   * @param  {[datarow]} row       DataRow instance to put under given DataSet ID
+   * @param  {[string]} dataSetId   Parent DataSet ID
+   * @return the original row passed in, with modifications.
+   */
+  _prepareDataRowInDataSet(row, dataSetId) {
+    _.each(row, (val,key) => {
+      if (typeof(val) == 'string') {
+        if (val.endsWith('%')) {
+          let what = Number.parseFloat(val)
+          if (_.isNumber(what) && !Number.isNaN(what)) {
+            val = what
+          }
+        }
+      } else if (typeof(val) == 'undefined') {
+        val = null
+      }
+      row[key] = val
+    })
+    row.dataset = dataSetId
+    //nb: must return changed row rather than new object
+    return row
+  }
+
+  /**
+   * Based on supplied data row, build array of field info objects
+   * which include a unique field-id for each property on the row.
+   * @param  {[datarow]} sample [representative row of data]
+   * @return { array of object}        [name, id, and type properties for each field]
+   */
+  _buildFieldInfo(sample) {
+    let fields = new Array()
+    _.each(_.keys(sample), (key,index) => {
+      let type = field.STRING_TYPE
+      let val = sample[key]
+      if (typeof(val) == 'string') {
+        if (val.endsWith('%')) {
+          type = field.PERCENT_TYPE
+        } else if (! isNaN(parseFloat(val))) {
+          type = field.DECIMAL_TYPE
+        }
+      }
+      if ( type != field.PERCENT_TYPE && 
+        ((-1 < key.toLowerCase().indexOf('percent')) || (-1 < key.indexOf('%'))) ) {
+        type = field.PERCENT_TYPE
+      }
+      if (! _.contains(["true", "id", "dataset", "createdAt", "updatedAt"], key)) {
+        let field = field.FIELD_DEFAULTS
+        field.name = key
+        field.id = this._generateUniqueId()
+        field.label = morph.toTitle(key)
+        field.type = type
+        fields.push(field)
+      }
+    })
+    return fields
+  }
+
+   /**
+    * override EditController save:
+   * handle POST response for edit existing 
+   * or create new dataset
+   */
+  save(req, res) {
+    let dataSetId = req.params.id
+    this.log.debug("save DataSet id: ", dataSetId)

```

Comment:
A note on style here: generally, I think the best practice is to do your checking for exceptions/validation at the top, then have the body of the method be the normal path.  So here, I might rewrite as:

```javascript
save(req, res) {
  return Promise.resolve().then(() => {
    if(!dataSetId && !req.body.name) throw new Error("Name is required")
    if(!dataSetId && req.body.name) return this.model.create({name: req.body.name, source: req.body.source, description: req.body.description})
    ....
  }).then((record) =>{
    this.log.debug("dataset created", record.id)
    return res.redirect(this.routePrefix)
  }).catch((e) => {
    req.flash('error', e)
    return res.redirect(this.routePrefix + '/create')
  })
})
```